src/http11.rs

```diff
@@ -0,0 +1,311 @@
+//! Adapts the HTTP/1.1 implementation into the `HttpMessage` API.
+use std::io::{self, Write, BufWriter, Read};
+use std::net::Shutdown;
+
+use method::{Method};
+use header::{ContentLength, TransferEncoding};
+use header::Encoding::Chunked;
+
+use net::{NetworkConnector, NetworkStream, ContextVerifier};
+use http::{HttpWriter, LINE_ENDING};
+use http::HttpReader::{SizedReader, ChunkedReader, EofReader};
+use http::HttpWriter::{ChunkedWriter, SizedWriter, EmptyWriter};
+use buffer::BufReader;
+use http::{self, HttpReader};
+
+use message::{
+    Protocol,
+    HttpMessage,
+    RequestHead,
+    ResponseHead,
+};
+use header;
+use version;
+
+/// An implementation of the `HttpMessage` trait for HTTP/1.1.
+#[derive(Debug)]
+pub struct Http11Message {
+    stream: Option<Box<NetworkStream + Send>>,
```

Comment:
All right, say there's an enum:

```rust
enum MessageState {
    Idle(Box<NetworkStream + Send>),
    Writing(HttpWriter<BufWriter<Box<NetworkStream + Send>>>),
    Reading(HttpReader<BufReader<Box<NetworkStream + Send>>>),
}
```

Now in `set_outgoing(&mut self, ...)`, the internal state of the message needs to transition to `MessageState::Writing` and the writer needs to be based on the stream originally wrapped in the `Idle` variant. I'm not sure how the `Box<NetworkStream>` should be moved into a different variant here.

`set_outgoing` takes `&mut self`, so neither the full state enum nor the `Box` within the `Idle` variant can be moved. The only thing I can think of around that would be to use `mem::replace`, but it'd still require something (akin to `None`) to be put in place of the original state while the stream is juggled into the correct Writer.

This `None`-like state, though, would still imply that it's technically possible to lose the stream along the way, so it's only a slight improvement...

Is there some better approach to implementing the transition which would avoid this `None` state altogether?